tls/s2n_ktls_io.c

```diff
@@ -319,7 +318,56 @@ ssize_t s2n_ktls_sendv_with_offset(struct s2n_connection *conn, const struct iov
     }
 
     size_t bytes_written = 0;
-    POSIX_GUARD_RESULT(s2n_ktls_sendmsg(conn, TLS_APPLICATION_DATA, bufs, count,
-            blocked, &bytes_written));
+    POSIX_GUARD_RESULT(s2n_ktls_sendmsg(conn->send_io_context, TLS_APPLICATION_DATA,
+            bufs, count, blocked, &bytes_written));
+    return bytes_written;
+}
+
+int s2n_ktls_send_cb(void *io_context, const uint8_t *buf, uint32_t len)
+{
+    /* For now, all control records are assumed to be alerts.
+     * We can set the record_type on the io_context in the future.
+     */
+    const uint8_t record_type = TLS_ALERT;
```

Comment:
I suspect I am missing something because I don't know why all the record types would be alerts right now.... Doesn't this function also send application data?

Reply:
This function doesn't send application data. Application data is special in ktls: because we just need to pass it as-is to the kernel, without any modification or encryption, we don't need to store it at all. So we don't use s2n_record_writev (because that just buffers the data into conn->out) and we don't use s2n_flush (because that just tries to write buffered data). We just call sendmsg directly. See https://github.com/aws/s2n-tls/pull/4174#discussion_r1310647133.

We don't need to encrypt control messages like alerts in ktls either, but we do need to store them so that we can handle partial writes and blocking. So we do use s2n_record_writev and s2n_flush.

Reply:
Hmmm although looking at that again, now that we're using s2n_flush for ktls, we should probably move the s2n_ktls_sendv_with_offset call to after the s2n_flush call. It doesn't matter right now, but it will matter if we support warning alerts or post-handshake messages.